pkg/scheduler/plugins/gang/gang_plugin.go

```diff
@@ -0,0 +1,507 @@
+		return subPrio1 > subPrio2
+	}
+
+	creationTime1 := p.GetCreatTime(podInfo1)
+	creationTime2 := p.GetCreatTime(podInfo2)
+	if creationTime1.Equal(creationTime2) {
+		return GetNamespaceSplicingName(podInfo1.Pod.Name, podInfo1.Pod.Namespace) < GetNamespaceSplicingName(podInfo2.Pod.Name, podInfo2.Pod.Namespace)
+	}
+	return creationTime1.Before(creationTime2)
+}
+
+// PreFilter
+//if non-strict-mode, we only do step1 and step2:
+// i.Check whether childes in Gang has met the requirements of minimum number under each Gang, and reject the pod if negative.
+// ii.Check whether the Gang has been timeout(check the pod's annotation,later introduced at Permit section), and reject the pod if positive.
+// iii.Check whether the Gang has met the scheduleCycleValid check, and reject the pod if negative.
+// iv.Try update scheduleCycle, scheduleCycleValid, childrenScheduleRoundMap as mentioned above.
+func (p *GangPlugin) PreFilter(ctx context.Context, state *framework.CycleState, pod *corev1.Pod) *framework.Status {
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		GetGangNameByPod(pod))
+	gang := p.gangCache.GetGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, "can't find gang")
+	}
+
+	//check if gang is timeout by pod's Annotation
+	if pod.Annotations[extension.AnnotationGangTimeout] == "true" {
+		klog.Errorf("gang has timeout, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, "gang has timeout")
+	}
+
+	if gang.GetChildrenNum() < gang.GetGangMinNum() {
+		klog.Errorf("gang child pod not collect enough, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, "gang child pod not collect enough")
+	}
+
+	if pod.Annotations[extension.AnnotationGangTimeout] == "true" {
+		klog.Errorf("gang has timeout, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, "gang has timeout")
+	}
+
+	gangMode := gang.GetGangMode()
+	if gangMode == extension.GangModeStrict {
+		defer gang.TryUpdateScheduleCycle()
+
+		podScheduleCycle := gang.GetChildScheduleCycle(pod)
+		gangScheduleCycle := gang.GetGangScheduleCycle()
+		if podScheduleCycle >= gangScheduleCycle {
+			klog.Errorf("pod's schedule cycle too large, gangName:%v, podName:%v, podCycle:%v, gangCycle:%v",
+				gangId, GetNamespaceSplicingName(pod.Namespace, pod.Name), podScheduleCycle, gangScheduleCycle)
+			return framework.NewStatus(framework.UnschedulableAndUnresolvable, "pod's schedule cycle too large")
+		}
+
+		gang.SetChildScheduleCycle(pod, gangScheduleCycle)
+
+		if !gang.IsGangScheduleCycleValid() {
```

Comment:
!gang.IsGangScheduleCycleValid() should first check. then check podScheduleCycle;  SetChildScheduleCycle and TryUpdateScheduleCycle both in defer